src/vos/vos_query.c

```diff
@@ -232,6 +229,111 @@ query_recx(struct open_query *query, daos_recx_t *recx)
 	return rc;
 }
 
+static int
+query_normal_recx(struct open_query *query, daos_recx_t *recx)
+{
+	struct evt_extent filter_ex;
+	int		  opc;
+
+	memset(recx, 0, sizeof(*recx));
+	/* query visible last recx */
+	opc = EVT_ITER_EMBEDDED | EVT_ITER_VISIBLE | EVT_ITER_SKIP_HOLES;
+	if (query->qt_flags & VOS_GET_MAX)
+		opc |= EVT_ITER_REVERSE;
+
+	filter_ex.ex_lo = 0;
+	filter_ex.ex_hi = ~(uint64_t)0;
+
+	return _query_recx(query, recx, &filter_ex, opc);
+}
+
+static int
+query_ec_normal_recx(struct open_query *query, daos_recx_t *recx,
+		     struct evt_extent *filter_ex,
+		     bool visible)
+{
+	struct evt_extent tmp_ex;
+	int		  opc;
+
+	memset(recx, 0, sizeof(*recx));
+	/* query visible last recx */
+	opc = EVT_ITER_EMBEDDED;
+	if (query->qt_flags & VOS_GET_MAX)
+		opc |= EVT_ITER_REVERSE;
+	if (visible)
+		opc |= EVT_ITER_VISIBLE | EVT_ITER_SKIP_HOLES;
+
+	if (filter_ex) {
+		tmp_ex = *filter_ex;
+	} else {
+		tmp_ex.ex_lo = 0;
+		tmp_ex.ex_hi = DAOS_EC_PARITY_BIT - 1;
+	}
+	D_DEBUG(DB_TRACE, "search opc %x ["DF_U64"-"DF_U64"]\n", opc,
+		tmp_ex.ex_lo, tmp_ex.ex_hi);
+	return _query_recx(query, recx, &tmp_ex, opc);
+}
+
+static int
+query_ec_parity_recx(struct open_query *query, daos_recx_t *recx)
+{
+	struct evt_extent filter_ex;
+	int		  opc;
+
+	memset(recx, 0, sizeof(*recx));
+	/* query visible last recx */
+	opc = EVT_ITER_EMBEDDED | EVT_ITER_VISIBLE | EVT_ITER_SKIP_HOLES;
+	if (query->qt_flags & VOS_GET_MAX)
+		opc |= EVT_ITER_REVERSE;
+
+	filter_ex.ex_lo = DAOS_EC_PARITY_BIT;
+	filter_ex.ex_hi = ~(uint64_t)0;
+
+	return _query_recx(query, recx, &filter_ex, opc);
+}
+
+static int
+query_recx(struct open_query *query, daos_recx_t *recxs)
+{
+	struct evt_extent filter_ex;
+	int		  rc;
+
+	if (!(query->qt_flags & VOS_GET_RECX_EC))
+		return query_normal_recx(query, &recxs[0]);
+
+	/* visible recxs */
+	rc = query_ec_normal_recx(query, &recxs[0], NULL, true);
+	if (rc && rc != -DER_NONEXIST)
+		return rc;
+
+	/* visible parity recxs */
+	rc = query_ec_parity_recx(query, &recxs[1]);
+	if (rc) {
+		if (rc == -DER_NONEXIST)
+			rc = 0;
+		return rc;
+	}
+
+	/* possible punched recxs within this parity recx */
+	filter_ex.ex_lo = recxs[1].rx_idx;
+	D_ASSERT((filter_ex.ex_lo & DAOS_EC_PARITY_BIT) != 0);
+	filter_ex.ex_lo &= ~DAOS_EC_PARITY_BIT;
+	D_ASSERT(filter_ex.ex_lo % query->qt_cell_size == 0);
+	filter_ex.ex_hi = (((filter_ex.ex_lo + recxs[1].rx_nr) / query->qt_cell_size) *
+			   query->qt_stripe_size);
+	filter_ex.ex_lo = ((filter_ex.ex_lo / query->qt_cell_size) *
+			   query->qt_stripe_size);
+
+	rc = query_ec_normal_recx(query, &recxs[2], &filter_ex, false);
+	if (rc == -DER_NONEXIST)
+		rc = 0;
+
+	D_DEBUG(DB_TRACE, "get punched recxs "DF_U64"/"DF_U64"\n", recxs[2].rx_idx,
```

Comment:
what if there is more than one contiguous punch?

Reply:
indeed, will update the patch.